fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java

```diff
@@ -38,37 +47,40 @@ public FDBDirectoryLockFactory(FDBDirectory directory) {
     }
 
     @Override
-    public Lock obtainLock(final Directory dir, final String lockName) throws IOException {
+    public Lock obtainLock(final Directory dir, final String lockName) {
         // dir is ignored
-        return new FDBDirectoryLock(directory, lockName);
+        return new FDBDirectoryLock(directory.getAgilityContext(), lockName, directory.fileLockKey(lockName));
     }
 
     private static class FDBDirectoryLock extends Lock {
 
-        final FDBDirectory directory;
+        final AgilityContext agilityContext;
         final String lockName;
         final long timeStampMillis;
+        final byte[] fileLockKey;
         boolean closed;
+        private static final Logger LOGGER = LoggerFactory.getLogger(FDBDirectoryLock.class);
 
-        public FDBDirectoryLock(final FDBDirectory directory, final String lockName) {
-            this.directory = directory;
-            this.lockName = lockName;
+        public FDBDirectoryLock(final AgilityContext agilityContext, final String lockName, byte[] fileLockKey) {
+            this.agilityContext = agilityContext;
+            this.lockName = lockName; // for log messages
+            this.fileLockKey = fileLockKey;
             this.timeStampMillis = System.currentTimeMillis();
-            directory.fileLockSet(lockName, timeStampMillis);
+            fileLockSet( timeStampMillis);
         }
 
         @Override
         public void close() {
-            directory.fileLockClear(lockName, timeStampMillis);
+            fileLockClear(timeStampMillis);
             closed = true;
         }
 
         @Override
-        public void ensureValid() throws IOException {
+        public void ensureValid() {
```

Comment:
IIUIC, Lucene is expecting a simple file as a lock. `SimpleFSLockFactory`, at least, does not use an FS lock on the file, hence `ensureValid()` is not expected to be atomic or to be performed under a lock.

Reply:
AFAICT `SimpleFSLockFactory` will never steal the lock, so, I think, we have an additional risk here.
I think making it transactional would be ideal, if we can (after a brief scan of the code, that looks tricky at best), so perhaps having `ensureValid()` check that the timestamp is no more than 30 minutes ago? Which should help ensure that another instance won't steal before the next time it does an operation (assuming clock skew is under 30 minutes).

Reply:
Do you mean the local timestamp? 
(I.e. `timeStampMillis`)
If so, why 30 minutes and not an hour (at least until we'll implement an heart bit)?

Reply:
Yes, I mean `timeStampMillis`.
Using 30 minutes provides some buffer for clock skew, and GC pauses and the operations between calling validateLock.
So if you have two instances: a and b, and b has a clock skew putting it 1 minute ahead you'll have:
t0: a creates a lock at t0
t53: a validates the lock, and proceeds to do something that takes 8 minutes before validating again
t59: b (which believes it is t60) tries to create a lock, and concludes that the current lock at t0 is one hour in the past, and should be stolen.
t61: a validates and sees that it no-longer has the lock, but at this point it wrote some data after b stole the lock


I'm not sure exactly what these two numbers should be, but 30 minutes seems like more than enough time to accomplish a merge in the short term, and gives plenty of buffer to avoid continuing to do work after another process has stolen the lock.

Looking at the usages, it looks like it validates the lock whenever it creates files or modifies the file system, and each of those operations on a proper filesystem would get flushed to the filesystem (although not to disk) when the method is called, whereas we might create a "file" and then continue to write to the file and do other things before it is actually committed.

Reply:
✅